compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -254,14 +271,18 @@ impl ResolverAstLoweringExt for ResolverAstLowering {
 
 /// Context of `impl Trait` in code, which determines whether it is allowed in an HIR subtree,
 /// and if so, what meaning it has.
-#[derive(Debug, Copy, Clone, PartialEq, Eq)]
-enum ImplTraitContext {
+#[derive(Debug)]
+enum ImplTraitContext<'b, 'itctx> {
     /// Treat `impl Trait` as shorthand for a new universal generic parameter.
     /// Example: `fn foo(x: impl Debug)`, where `impl Debug` is conceptually
     /// equivalent to a fresh universal parameter like `fn foo<T: Debug>(x: T)`.
     ///
     /// Newly generated parameters should be inserted into the given `Vec`.
-    Universal,
+    Universal {
+        apit_nodes: &'b mut Vec<&'itctx Ty>,
+    },
+
+    UniversalInRPIT,
```

Comment:
What is the difference with `Universal`?

Reply:
The difference is that `Universal` is used when lowering APITs and it carries `apit_nodes` vector which is filled as APITs are being lowered. `UniversalInRPIT` is used when lowering universal inside RPITs.
This is not great to be honest, we should probably want to do the same I've done for lifetimes, adding visitors and gather these apit_nodes on a side track.
I guess it would be better if we leave this for a followup PR.